stubs/SQLAlchemy/sqlalchemy/engine/base.pyi

```diff
@@ -82,88 +118,115 @@ class ExceptionContextImpl(ExceptionContext):
-        def __enter__(self): ...
-        def __exit__(self, type_, value, traceback) -> None: ...
-    def begin(self, close_with_result: bool = ...): ...
-    def transaction(self, callable_, *args, **kwargs): ...
-    def run_callable(self, callable_, *args, **kwargs): ...
-    def execute(self, statement, *multiparams, **params): ...
-    def scalar(self, statement, *multiparams, **params): ...
-    def connect(self, close_with_result: bool = ...): ...  # type: ignore[override]
-    def table_names(self, schema: Any | None = ..., connection: Any | None = ...): ...
-    def has_table(self, table_name, schema: Any | None = ...): ...
-    def raw_connection(self, _connection: Any | None = ...): ...
+        conn: Connection
+        transaction: Transaction
+        close_with_result: bool
+        def __init__(self, conn: Connection, transaction: Transaction, close_with_result: bool) -> None: ...
+        def __enter__(self) -> Connection: ...
+        def __exit__(
+            self, type_: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
+        ) -> None: ...
+    def begin(self, close_with_result: bool = ...) -> _trans_ctx: ...
+    # TODO:
+    # def transaction(self, callable_: Callable[Concatenate[Connection, _P], _T], *args: _P.args, **kwargs: _P.kwargs) -> _T | None: ...
+    def transaction(self, callable_: Callable[..., _T], *args: Any, **kwargs: Any) -> _T | None: ...
+    # TODO:
+    # def run_callable(self, callable_: Callable[Concatenate[Connection, _P], _T], *args: _P.args, **kwargs: _P.kwargs) -> _T: ...
+    def run_callable(self, callable_: Callable[..., _T], *args: Any, **kwargs: Any) -> _T: ...
+    @overload  # type: ignore[override]
+    def execute(
+        self,
+        statement: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
+        *multiparams: Mapping[str, Any],
+        **params: Any,
+    ) -> CursorResult: ...
+    @overload
+    def execute(self, statement: str, *multiparams: Any | tuple[Any, ...] | Mapping[str, Any], **params: Any) -> CursorResult: ...
+    @overload  # type: ignore[override]
+    def scalar(
+        self,
+        statement: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
+        *multiparams: Mapping[str, Any],
+        **params: Any,
+    ) -> Any: ...
+    @overload
+    def scalar(self, statement: str, *multiparams: Any | tuple[Any, ...] | Mapping[str, Any], **params: Any) -> Any: ...
+    def connect(self, close_with_result: bool = ...) -> Connection: ...  # type: ignore[override]
+    def table_names(self, schema: Any | None = ..., connection: Connection | None = ...): ...
+    def has_table(self, table_name: str, schema: Any | None = ...) -> bool: ...
+    def raw_connection(self, _connection: Connection | None = ...) -> DBAPIConnection: ...
 
 class OptionEngineMixin:
-    url: Any
-    dialect: Any
-    logging_name: Any
-    echo: Any
-    hide_parameters: Any
+    url: str
```

Comment:
How do you know that this is not a `Url` object?

Reply:
Actually looking at the code and trying it out, `Engine` uses an `URL` object, so I'm changing the instances accordingly.